### ISSUE TITLE
python313: unreproducible `libpython3.13.so.1.0` and `libpython3.13.a`

### ISSUE THREAD

**Issue by raboof** [NixOS/nixpkgs#346119]
Building this package multiple times does not yield bit-by-bit identical
results, complicating the detection of Continuous Integration (CI) breaches. For
more information on this issue, visit
[reproducible-builds.org](https://reproducible-builds.org/).

Fixing bit-by-bit reproducibility also has additional advantages, such as
avoiding hard-to-reproduce bugs, making content-addressed storage more effective
and reducing rebuilds in such systems.

### Steps To Reproduce

In the following steps, replace `<package>` with the canonical name of the
package.

#### 1. Build the package

This step will build the package. Specific arguments are passed to the command
to keep the build artifacts so we can compare them in case of differences.

Execute the following command:

```
nix-build '<nixpkgs>' -A python313 && nix-build '<nixpkgs>' -A python313 --check --keep-failed
```

Or using the new command line style:

```
nix build nixpkgs#python313 && nix build nixpkgs#python313 --rebuild --keep-failed
```

#### 2. Compare the build artifacts

If the previous command completes successfully, no differences were found and
there's nothing to do, builds are reproducible.
If it terminates with the error message `error: derivation '<X>' may not be
deterministic: output '<Y>' differs from '<Z>'`, use `diffoscope` to investigate
the discrepancies between the two build outputs. You may need to add the
`--exclude-directory-metadata recursive` option to ignore files and directories
metadata (*e.g. timestamp*) differences.

```
nix run nixpkgs#diffoscopeMinimal -- --exclude-directory-metadata recursive <Y> <Z>
```

#### 3. Examine the build log

To examine the build log, use:

```
nix-store --read-log $(nix-instantiate '<nixpkgs>' -A python313)
```

Or with the new command line style:

```
nix log $(nix path-info --derivation nixpkgs#python313)
```

### Additional context

Notably the `GNU NT_GNU_BUILD_ID` of `libpython3.13.so.1.0` is different - I though that was determined based on path, so that's surprising and something to double-check.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by hacscred** [comment]
Any particular reason for not including all of the changes you already found? Too much output?

**Comment by raboof** [comment]
Not really, it's easy enough to reproduce with the instructions above, and I figured picking out the difference that struck me most (the different `NT_GNU_BUILD_ID`) would be more valuable than pasting the whole thing. But here's a bit more:

<details>
<summary>some diffoscope output snippets</summary>
```
--- /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2
+++ /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2.check
│   --- /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2/lib
├── +++ /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2.check/lib
│ │   --- /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2/lib/libpython3.13.so.1.0
│ ├── +++ /nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2.check/lib/libpython3.13.so.1.0
│ │┄ File has been modified after NT_GNU_BUILD_ID has been applied.
│ │ ├── readelf --wide --symbols {}
│ │ │ @@ -1178,15 +1178,15 @@
│ │ │    1174: 00000000000c95f0   122 FUNC    GLOBAL DEFAULT   14 PyException_SetArgs
│ │ │    1175: 00000000002fac70     8 FUNC    GLOBAL DEFAULT   14 PyType_FromMetaclass
│ │ │    1176: 00000000001abde0   234 FUNC    GLOBAL DEFAULT   14 PyNumber_InPlaceRshift
│ │ │    1177: 00000000001cea00  1048 FUNC    GLOBAL DEFAULT   14 _PyXI_ApplyCapturedException
│ │ │    1178: 0000000000198e70   126 FUNC    GLOBAL DEFAULT   14 PySequence_Size
│ │ │    1179: 00000000001db100   710 FUNC    GLOBAL DEFAULT   14 PyObject_GetItem
│ │ │    1180: 00000000005b1ce0   416 OBJECT  GLOBAL DEFAULT   26 PyDictValues_Type
│ │ │ +  1181: 0000000000468d80   256 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_toupper
│ │ │ -  1181: 0000000000468da0   256 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_toupper
│ │ │    1182: 00000000005b83e0   416 OBJECT  GLOBAL DEFAULT   26 PyBytesIter_Type
│ │ │    1183: 00000000000e01a0   174 FUNC    GLOBAL DEFAULT   14 PyErr_SetHandledException
│ │ │    1184: 0000000000399980   382 FUNC    GLOBAL DEFAULT   14 _PyXI_EndInterpreter
│ │ │    1185: 000000000019a450   216 FUNC    GLOBAL DEFAULT   14 PyObject_GetIter
│ │ │    1186: 00000000002272f0   226 FUNC    GLOBAL DEFAULT   14 _PyObject_CallFunction_SizeT
│ │ │    1187: 00000000002c3ca0    10 FUNC    GLOBAL DEFAULT   14 _Py_open
│ │ │    1188: 0000000000354b90   287 FUNC    GLOBAL DEFAULT   14 PyEval_SetTraceAllThreads
│ │ │ @@ -1468,15 +1468,15 @@
│ │ │    1464: 00000000000f08f0   326 FUNC    GLOBAL DEFAULT   14 _PyObject_SetManagedDict
│ │ │    1465: 00000000005afcc0   416 OBJECT  GLOBAL DEFAULT   26 PyRange_Type
│ │ │    1466: 00000000003485e0    27 FUNC    GLOBAL DEFAULT   14 _PySemaphore_Wakeup
│ │ │    1467: 00000000002fb1c0     7 FUNC    GLOBAL DEFAULT   14 PyStructSequence_NewType
│ │ │    1468: 00000000000deaf0    73 FUNC    GLOBAL DEFAULT   14 _PyUnicode_IsNumeric
│ │ │    1469: 00000000000e7100    35 FUNC    GLOBAL DEFAULT   14 PyThread_exit_thread
│ │ │    1470: 00000000003500d0   216 FUNC    GLOBAL DEFAULT   14 _PyRecursiveMutex_Lock
│ │ │ +  1471: 0000000000468c80   256 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_tolower
│ │ │ -  1471: 0000000000468ca0   256 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_tolower
│ │ │    1472: 00000000000a4b00    87 FUNC    GLOBAL DEFAULT   14 PyInit__imp
│ │ │    1473: 000000000016cb80   968 FUNC    GLOBAL DEFAULT   14 _PyLong_AsByteArray
│ │ │    1474: 00000000000ca4e0    41 FUNC    GLOBAL DEFAULT   14 _Py_ResurrectReference
│ │ │    1475: 00000000000e6e20   107 FUNC    GLOBAL DEFAULT   14 PyThread_init_thread
│ │ │    1476: 00000000001b9c70    96 FUNC    GLOBAL DEFAULT   14 PyUnicode_Count
│ │ │    1477: 00000000002a7e00   898 FUNC    GLOBAL DEFAULT   14 PyErr_SetImportError
│ │ │    1478: 00000000000df8d0   183 FUNC    GLOBAL DEFAULT   14 _PyXI_ClearExcInfo
│ │ │ @@ -1769,15 +1769,15 @@
│ │ │    1765: 00000000002e2690   187 FUNC    GLOBAL DEFAULT   14 PyErr_WarnEx
│ │ │    1766: 00000000002fac80    22 FUNC    GLOBAL DEFAULT   14 PyType_FromModuleAndSpec
│ │ │    1767: 00000000000e0f60   276 FUNC    GLOBAL DEFAULT   14 _Py_hashtable_set
│ │ │    1768: 00000000002c11a0     9 FUNC    GLOBAL DEFAULT   14 PyUnicode_AsUTF16String
│ │ │    1769: 000000000034de30   351 FUNC    GLOBAL DEFAULT   14 PyMutex_Unlock
│ │ │    1770: 00000000005b7b60   416 OBJECT  GLOBAL DEFAULT   26 PyMethod_Type
│ │ │    1771: 00000000005a9e80   416 OBJECT  GLOBAL DEFAULT   26 PyContextVar_Type
│ │ │ +  1772: 000000000048e680   128 OBJECT  GLOBAL DEFAULT   16 _Py_ascii_whitespace
│ │ │ -  1772: 000000000048e6a0   128 OBJECT  GLOBAL DEFAULT   16 _Py_ascii_whitespace
│ │ │    1773: 00000000001908e0   431 FUNC    GLOBAL DEFAULT   14 PyLong_AsUnsignedLongLongMask
│ │ │    1774: 000000000012f770   345 FUNC    GLOBAL DEFAULT   14 PySet_Discard
│ │ │    1775: 00000000001ceef0   372 FUNC    GLOBAL DEFAULT   14 PyDescr_NewGetSet
│ │ │    1776: 00000000002eff90   339 FUNC    GLOBAL DEFAULT   14 PyFile_OpenCode
│ │ │    1777: 00000000000a4c80    87 FUNC    GLOBAL DEFAULT   14 PyInit__abc
│ │ │    1778: 00000000000e7720  1213 FUNC    GLOBAL DEFAULT   14 PyThread_acquire_lock_timed
│ │ │    1779: 0000000000085980 50900 FUNC    GLOBAL DEFAULT   14 _PyEval_EvalFrameDefault
│ │ │ @@ -1802,15 +1802,15 @@
│ │ │    1798: 00000000000e59a0    19 FUNC    GLOBAL DEFAULT   14 _PyInterpreterState_IsRunningMain
│ │ │    1799: 0000000000109dd0    61 FUNC    GLOBAL DEFAULT   14 PyInterpreterState_Get
│ │ │    1800: 00000000003aea80   140 FUNC    GLOBAL DEFAULT   14 PyImport_ExecCodeModule
│ │ │    1801: 00000000001d7460   322 FUNC    GLOBAL DEFAULT   14 PyObject_Vectorcall
│ │ │    1802: 00000000005351e0     8 OBJECT  GLOBAL DEFAULT   23 PyStructSequence_UnnamedField
│ │ │    1803: 00000000002c6c10   185 FUNC    GLOBAL DEFAULT   14 PyCodec_Encode
│ │ │    1804: 00000000000c9eb0   139 FUNC    GLOBAL DEFAULT   14 _PyDict_SizeOf
│ │ │ +  1805: 000000000046d178     8 OBJECT  GLOBAL DEFAULT   16 Py_Version
│ │ │ -  1805: 000000000046d198     8 OBJECT  GLOBAL DEFAULT   16 Py_Version
│ │ │    1806: 00000000005fe010    24 OBJECT  GLOBAL DEFAULT   28 _Py_HashSecret
│ │ │    1807: 00000000005acd00   416 OBJECT  GLOBAL DEFAULT   26 PyType_Type
│ │ │    1808: 00000000002ca0d0   159 FUNC    GLOBAL DEFAULT   14 PyErr_ProgramTextObject
│ │ │    1809: 0000000000174900   169 FUNC    GLOBAL DEFAULT   14 _PyBytesWriter_Prepare
│ │ │    1810: 00000000001366e0    92 FUNC    GLOBAL DEFAULT   14 PyErr_SetNone
│ │ │    1811: 00000000000c94b0   122 FUNC    GLOBAL DEFAULT   14 PyException_SetCause
│ │ │    1812: 000000000021bff0   276 FUNC    GLOBAL DEFAULT   14 _Py_dup
│ │ │ @@ -1889,15 +1889,15 @@
│ │ │    1885: 00000000001d8820   183 FUNC    GLOBAL DEFAULT   14 PyObject_CallOneArg
│ │ │    1886: 0000000000232280   101 FUNC    GLOBAL DEFAULT   14 PyObject_ClearWeakRefs
│ │ │    1887: 0000000000107470   211 FUNC    GLOBAL DEFAULT   14 PyFile_NewStdPrinter
│ │ │    1888: 00000000000c9980   107 FUNC    GLOBAL DEFAULT   14 _PyFloat_ExactDealloc
│ │ │    1889: 00000000000dfa30    45 FUNC    GLOBAL DEFAULT   14 PyErr_GetRaisedException
│ │ │    1890: 0000000000150e50   577 FUNC    GLOBAL DEFAULT   14 PyVectorcall_Call
│ │ │    1891: 0000000000357bc0    97 FUNC    GLOBAL DEFAULT   14 _PyMem_GetCurrentAllocatorName
│ │ │ +  1892: 0000000000464bd8     8 OBJECT  GLOBAL DEFAULT   16 PY_TIMEOUT_MAX
│ │ │ -  1892: 0000000000464bf8     8 OBJECT  GLOBAL DEFAULT   16 PY_TIMEOUT_MAX
│ │ │    1893: 0000000000140330   764 FUNC    GLOBAL DEFAULT   14 PyContextVar_Set
│ │ │    1894: 00000000001153b0   104 FUNC    GLOBAL DEFAULT   14 PyType_GetQualName
│ │ │    1895: 00000000005b8240   416 OBJECT  GLOBAL DEFAULT   26 PyBytes_Type
│ │ │    1896: 00000000000ebe20  2499 FUNC    GLOBAL DEFAULT   14 _PyUnicode_CheckConsistency
│ │ │    1897: 00000000000e8b10   109 FUNC    GLOBAL DEFAULT   14 _Py_write_noraise
│ │ │    1898: 00000000000c9b40    41 FUNC    GLOBAL DEFAULT   14 PyFrame_GetGlobals
│ │ │    1899: 000000000011f7d0  4265 FUNC    GLOBAL DEFAULT   14 PyUnicode_FromFormatV
│ │ │ @@ -1937,15 +1937,15 @@
│ │ │    1933: 00000000000ca190    70 FUNC    GLOBAL DEFAULT   14 PyModuleDef_Init
│ │ │    1934: 00000000001d58f0    97 FUNC    GLOBAL DEFAULT   14 _PyMonitoring_FirePyResumeEvent
│ │ │    1935: 00000000002e2130   101 FUNC    GLOBAL DEFAULT   14 PyModule_Create2
│ │ │    1936: 00000000000962b0    23 FUNC    GLOBAL DEFAULT   14 PyMem_RawFree
│ │ │    1937: 00000000000d7d00   229 FUNC    GLOBAL DEFAULT   14 _PyUnicode_Equal
│ │ │    1938: 00000000000c8870    11 FUNC    GLOBAL DEFAULT   14 PyVectorcall_NARGS
│ │ │    1939: 000000000018ceb0   336 FUNC    GLOBAL DEFAULT   14 PyLong_AsLongAndOverflow
│ │ │ +  1940: 0000000000468880  1024 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_table
│ │ │ -  1940: 00000000004688a0  1024 OBJECT  GLOBAL DEFAULT   16 _Py_ctype_table
│ │ │    1941: 00000000000c8f10    48 FUNC    GLOBAL DEFAULT   14 _Py_c_neg
│ │ │    1942: 00000000001d4b30   435 FUNC    GLOBAL DEFAULT   14 _PyMonitoring_FireStopIterationEvent
│ │ │    1943: 00000000000d76f0    40 FUNC    GLOBAL DEFAULT   14 PyType_GetTypeDataSize
│ │ │    1944: 00000000000d8b20 20067 FUNC    GLOBAL DEFAULT   14 _PyUnicode_ToNumeric
│ │ │    1945: 00000000000e5a90     7 FUNC    GLOBAL DEFAULT   14 PyThreadState_Next
│ │ │    1946: 000000000012a360   324 FUNC    GLOBAL DEFAULT   14 PyObject_GetOptionalAttrString
│ │ │    1947: 00000000005a0cb0     8 OBJECT  GLOBAL DEFAULT   26 PyExc_TimeoutError
│ │ │ @@ -2287,15 +2287,15 @@
│ │ │     219: 0000000000098960    75 FUNC    LOCAL  DEFAULT   14 zip_traverse
│ │ │     220: 00000000000989b0   134 FUNC    LOCAL  DEFAULT   14 _PyFrame_PushUnchecked
│ │ │     221: 0000000000098a40   409 FUNC    LOCAL  DEFAULT   14 get_standard_encoding
```
(...)
```
│ │ │   10151: 00000000005fe000     8 OBJECT  GLOBAL DEFAULT   28 PyImport_FrozenModules
│ │ │   10152: 0000000000123700   444 FUNC    GLOBAL DEFAULT   14 PyGen_NewWithQualName
│ │ │   10153: 00000000002f1020   700 FUNC    GLOBAL DEFAULT   14 PyCapsule_Import
│ │ │   10154: 00000000000df2c0   163 FUNC    GLOBAL DEFAULT   14 PyEval_GetFuncDesc
│ │ │ + 10155: 0000000000464bd8     8 OBJECT  GLOBAL DEFAULT   16 PY_TIMEOUT_MAX
│ │ │ - 10155: 0000000000464bf8     8 OBJECT  GLOBAL DEFAULT   16 PY_TIMEOUT_MAX
│ │ │   10156: 000000000034f350   108 FUNC    GLOBAL DEFAULT   14 PyObject_GetArenaAllocator
│ │ │   10157: 0000000000000000     0 OBJECT  GLOBAL DEFAULT  UND stdout@GLIBC_2.2.5
│ │ │   10158: 00000000000deaa0    73 FUNC    GLOBAL DEFAULT   14 _PyUnicode_IsDigit
│ │ │   10159: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND readdir64@GLIBC_2.2.5
│ │ │   10160: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND closefrom@GLIBC_2.34
│ │ │   10161: 000000000025de30   223 FUNC    GLOBAL DEFAULT   14 PySequence_List
│ │ │   10162: 00000000003cabd0   144 FUNC    GLOBAL DEFAULT   14 PyErr_SetInterrupt
│ │ ├── readelf --wide --relocs {}
│ │ │ @@ -8,65 +8,65 @@
│ │ │  0000000000530a40  0000000000000008 R_X86_64_RELATIVE                         49e99d
│ │ │  0000000000530a48  0000000000000008 R_X86_64_RELATIVE                         3f04a0
│ │ │  0000000000530a58  0000000000000008 R_X86_64_RELATIVE                         4a715e
│ │ │  0000000000530a60  0000000000000008 R_X86_64_RELATIVE                         3e07a0
│ │ │  0000000000530a70  0000000000000008 R_X86_64_RELATIVE                         4a69f2
│ │ │  0000000000530a78  0000000000000008 R_X86_64_RELATIVE                         3d9f00
│ │ │  0000000000530aa0  0000000000000008 R_X86_64_RELATIVE                         4a7186
│ │ │ +0000000000530aa8  0000000000000008 R_X86_64_RELATIVE                         43d120
│ │ │ -0000000000530aa8  0000000000000008 R_X86_64_RELATIVE                         43d140
│ │ │  0000000000530ab8  0000000000000008 R_X86_64_RELATIVE                         4a81ef
│ │ │ +0000000000530ac0  0000000000000008 R_X86_64_RELATIVE                         4332e0
│ │ │ -0000000000530ac0  0000000000000008 R_X86_64_RELATIVE                         433300
│ │ │  0000000000530ad0  0000000000000008 R_X86_64_RELATIVE                         4a3c96
│ │ │ +0000000000530ad8  0000000000000008 R_X86_64_RELATIVE                         432240
│ │ │ -0000000000530ad8  0000000000000008 R_X86_64_RELATIVE                         432260
│ │ │  0000000000530ae8  0000000000000008 R_X86_64_RELATIVE                         4a7179
│ │ │ +0000000000530af0  0000000000000008 R_X86_64_RELATIVE                         426c00
│ │ │ -0000000000530af0  0000000000000008 R_X86_64_RELATIVE                         426c20
│ │ │  0000000000530b00  0000000000000008 R_X86_64_RELATIVE                         4a718a
│ │ │ +0000000000530b08  0000000000000008 R_X86_64_RELATIVE                         4258e0
│ │ │ -0000000000530b08  0000000000000008 R_X86_64_RELATIVE                         425900
│ │ │  0000000000530b18  0000000000000008 R_X86_64_RELATIVE                         4a7198
│ │ │ +0000000000530b20  0000000000000008 R_X86_64_RELATIVE                         423d20
│ │ │ -0000000000530b20  0000000000000008 R_X86_64_RELATIVE                         423d40
│ │ │  0000000000530b30  0000000000000008 R_X86_64_RELATIVE                         4a71a4
│ │ │ +0000000000530b38  0000000000000008 R_X86_64_RELATIVE                         41d100
│ │ │ -0000000000530b38  0000000000000008 R_X86_64_RELATIVE                         41d120

```
(...)
```
│ │ │  00000000005a4278  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ │  00000000005a4988  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ │  00000000005a4a28  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ │  00000000005a66d8  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ │  00000000005a6a68  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ │  00000000005aa6f0  0000029f00000001 R_X86_64_64            0000000000100550 PyType_GenericAlloc + 0
│ │ ├── readelf --wide --notes {}
│ │ │ @@ -1,8 +1,8 @@
│ │ │  
│ │ │  Displaying notes found in: .note.gnu.property
│ │ │    Owner                Data size 	Description
│ │ │    GNU                  0x00000020	NT_GNU_PROPERTY_TYPE_0	      Properties: x86 feature used: x86, XMM, x86 ISA used: x86-64-baseline
│ │ │  
│ │ │  Displaying notes found in: .note.gnu.build-id
│ │ │    Owner                Data size 	Description
│ │ │ +  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)	    Build ID: 6796c7c2a5cd988df47860a9cdcee340b2a70f88
│ │ │ -  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)	    Build ID: ab33e97ea306280645396ae81303d9613e83b5f8
│ │ ├── strings --all --bytes=8 {}
│ │ │ @@ -8312,15 +8312,15 @@
│ │ │  MAXLINESr-
│ │ │  _Printer.__repr__<
│ │ │  Nz0Hit Return for more, or q (and Return) to quit: r
│ │ │  IndexError)
```
(...)
</details>

**Comment by raboof** [comment]
Thanks for sharing your input!
 
> The only reason for that ID to be modified, is if it's modified at the wrong time. So, the way the build system should work is that it generates that ID as the last step (and nothing should touch it afterwards). Clearly, that's not what is happening.

Why do we care if it's touched afterwards? As long as the end-result is deterministic that should be fine - though I agree the nicest way is to just generate it deterministically from the start and not touch it anymore after that.

> So, one way to debug this is to strace the complete build process and track exactly when the build ID is set (this is done via a flag to ld or the compiler). Then, you should get a complete history of.rhe build process and presumably some program is running after it which touches that file again, which strace can also tell you. Combine those steps and you can solve the problem, if one exists.

stracing all of the build indeed seems rather brute-force - I don't think that would be easy to achieve, and even if we could it would produce an unusable amount of output.

Looking at the 'regular' logs, it seems libpython3.13.so.1.0 is created with

```
gcc -shared -L/nix/store/qhf8n3srrg0vdd44k7q71fb0p5az15b5-bzip2-1.0.8/lib -L/nix/store/f0dmadrh535bpayz2r4q5lslg0m17lnw-expat-2.6.2/lib -L/nix/store/4ybgqxv43z9sk9lccwq6dgmz6j32syr1-libffi-3.4.6/lib -L/nix/store/lhfgwjwwdacygiq702yvbahvpqmxivpx-libxcrypt-4.4.36/lib -L/nix/store/3hxxbbjc8r66nravvjind6ixhz7cpij1-mpdecimal-4.0.0/lib -L/nix/store/z7nr6aqlzv51pk5ar8bgzg2alfqvi8fd-ncurses-6.4.20221231/lib -L/nix/store/1w90l4fm5lzhlybipfilyjij2das6w98-openssl-3.0.14/lib -L/nix/store/g2i5j8i0bwfih7hg2wbc15k2969bckn3-sqlite-3.46.0/lib -L/nix/store/4w3c2c11j2kmwxrznji2i7is53z5aldi-xz-5.6.2/lib -L/nix/store/rqs1zrcncqz3966khjndg1183cpdnqxs-zlib-1.3.1/lib -L/nix/store/897xqnq52vw76991r5m80h9j91370vj9-tzdata-2024a/lib -L/nix/store/8ad0k7rbn3rr9qiya29464g52asndxqi-gdbm-1.24-lib/lib -L/nix/store/xxcf5gwyn5pldv4b4wa2jw6vqg7v55y6-readline-8.2p10/lib -L/nix/store/qhf8n3srrg0vdd44k7q71fb0p5az15b5-bzip2-1.0.8/lib -L/nix/store/f0dmadrh535bpayz2r4q5lslg0m17lnw-expat-2.6.2/lib -L/nix/store/4ybgqxv43z9sk9lccwq6dgmz6j32syr1-libffi-3.4.6/lib -L/nix/store/lhfgwjwwdacygiq702yvbahvpqmxivpx-libxcrypt-4.4.36/lib -L/nix/store/3hxxbbjc8r66nravvjind6ixhz7cpij1-mpdecimal-4.0.0/lib -L/nix/store/z7nr6aqlzv51pk5ar8bgzg2alfqvi8fd-ncurses-6.4.20221231/lib -L/nix/store/1w90l4fm5lzhlybipfilyjij2das6w98-openssl-3.0.14/lib -L/nix/store/g2i5j8i0bwfih7hg2wbc15k2969bckn3-sqlite-3.46.0/lib -L/nix/store/4w3c2c11j2kmwxrznji2i7is53z5aldi-xz-5.6.2/lib -L/nix/store/rqs1zrcncqz3966khjndg1183cpdnqxs-zlib-1.3.1/lib -L/nix/store/897xqnq52vw76991r5m80h9j91370vj9-tzdata-2024a/lib -L/nix/store/8ad0k7rbn3rr9qiya29464g52asndxqi-gdbm-1.24-lib/lib -L/nix/store/xxcf5gwyn5pldv4b4wa2jw6vqg7v55y6-readline-8.2p10/lib -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g  -Wl,-hlibpython3.13.so.1.0 -o libpython3.13.so.1.0 Modules/getbuildinfo.o Parser/token.o  Parser/pegen.o Parser/pegen_errors.o Parser/action_helpers.o Parser/parser.o Parser/string_parser.o Parser/peg_api.o Parser/lexer/buffer.o Parser/lexer/lexer.o Parser/lexer/state.o Parser/tokenizer/file_tokenizer.o Parser/tokenizer/readline_tokenizer.o Parser/tokenizer/string_tokenizer.o Parser/tokenizer/utf8_tokenizer.o Parser/tokenizer/helpers.o Parser/myreadline.o Objects/abstract.o Objects/boolobject.o Objects/bytes_methods.o Objects/bytearrayobject.o Objects/bytesobject.o Objects/call.o Objects/capsule.o Objects/cellobject.o Objects/classobject.o Objects/codeobject.o Objects/complexobject.o Objects/descrobject.o Objects/enumobject.o Objects/exceptions.o Objects/genericaliasobject.o Objects/genobject.o Objects/fileobject.o Objects/floatobject.o Objects/frameobject.o Objects/funcobject.o Objects/iterobject.o Objects/listobject.o Objects/longobject.o Objects/dictobject.o Objects/odictobject.o Objects/memoryobject.o Objects/methodobject.o Objects/moduleobject.o Objects/namespaceobject.o Objects/object.o Objects/obmalloc.o Objects/picklebufobject.o Objects/rangeobject.o Objects/setobject.o Objects/sliceobject.o Objects/structseq.o Objects/tupleobject.o Objects/typeobject.o Objects/typevarobject.o Objects/unicodeobject.o Objects/unicodectype.o Objects/unionobject.o Objects/weakrefobject.o Python/asm_trampoline.o Python/_warnings.o Python/Python-ast.o Python/Python-tokenize.o Python/asdl.o Python/assemble.o Python/ast.o Python/ast_opt.o Python/ast_unparse.o Python/bltinmodule.o Python/brc.o Python/ceval.o Python/codecs.o Python/compile.o Python/context.o Python/critical_section.o Python/crossinterp.o Python/dynamic_annotations.o Python/errors.o Python/flowgraph.o Python/frame.o Python/frozenmain.o Python/future.o Python/gc.o Python/gc_free_threading.o Python/gc_gil.o Python/getargs.o Python/getcompiler.o Python/getcopyright.o Python/getplatform.o Python/getversion.o Python/ceval_gil.o Python/hamt.o Python/hashtable.o Python/import.o Python/importdl.o Python/initconfig.o Python/interpconfig.o Python/instrumentation.o Python/instruction_sequence.o Python/intrinsics.o Python/jit.o Python/legacy_tracing.o Python/lock.o Python/marshal.o Python/modsupport.o Python/mysnprintf.o Python/mystrtoul.o Python/object_stack.o Python/optimizer.o Python/optimizer_analysis.o Python/optimizer_symbols.o Python/parking_lot.o Python/pathconfig.o Python/preconfig.o Python/pyarena.o Python/pyctype.o Python/pyfpe.o Python/pyhash.o Python/pylifecycle.o Python/pymath.o Python/pystate.o Python/pythonrun.o Python/pytime.o Python/qsbr.o Python/bootstrap_hash.o Python/specialize.o Python/structmember.o Python/symtable.o Python/sysmodule.o Python/thread.o Python/traceback.o Python/tracemalloc.o Python/getopt.o Python/pystrcmp.o Python/pystrtod.o Python/pystrhex.o Python/dtoa.o Python/formatter_unicode.o Python/fileutils.o Python/suggestions.o Python/perf_trampoline.o Python/perf_jit_trampoline.o Python/dynload_shlib.o     Modules/config.o Modules/main.o Modules/gcmodule.o Modules/atexitmodule.o  Modules/faulthandler.o  Modules/posixmodule.o  Modules/signalmodule.o  Modules/_tracemalloc.o  Modules/_suggestions.o  Modules/_codecsmodule.o  Modules/_collectionsmodule.o  Modules/errnomodule.o  Modules/_io/_iomodule.o Modules/_io/iobase.o Modules/_io/fileio.o Modules/_io/bytesio.o Modules/_io/bufferedio.o Modules/_io/textio.o Modules/_io/stringio.o  Modules/itertoolsmodule.o  Modules/_sre/sre.o  Modules/_sysconfig.o  Modules/_threadmodule.o  Modules/timemodule.o  Modules/_typingmodule.o  Modules/_weakref.o  Modules/_abc.o  Modules/_functoolsmodule.o  Modules/_localemodule.o  Modules/_operator.o  Modules/_stat.o  Modules/symtablemodule.o  Modules/pwdmodule.o Modules/getpath.o Python/frozen.o                          -ldl -L/nix/store/lhfgwjwwdacygiq702yvbahvpqmxivpx-libxcrypt-4.4.36/lib  -lm
```

... so the build id is not specified explicitly there. This line is no different between the hydra and local build logs, though - in fact there don't seem to be any real differences except for ordering and the parallelism (2 vs 16). Interestingly my local rebuild appears to create `/nix/store/a3njanrj2adfn0pz2hmh9i7vxkk36qfr-python3-3.13.0rc2/lib/python3.13` twice, but that seems harmless as well. I'll test with parallelism 2 just to be sure.

> it might just be an interaction of NixOS defaults

Possibly!

> assuming it is already reproducible on other operating systems (if those other systems have patches, first look at those).

That would indeed be interesting to investigate.

> I didn't see you using --pure flags, but perhaps those are implicit in flakes. I would not use flakes in bug reports, since they are still experimental.

I did not use flakes in my test, I just used `nix-build`. Both `nix-build` and `nix build` are pure by default, though - `--pure` is only a thing for `nix-shell`.

> If I had to wildly speculate, I would guess patchelf is executed after the build ID has been set, but if that were the case, perhaps patchelf should be modified to include an assertion or to log a warning when modifying the cryptographic hashes of other build systems, like this ID.

I think it's very likely patchelf is executed after the build ID has been set, but I don't see how that is relevant: AFAIK the build id is not a cryptographic hash, but just an ID, and should not be touched by patchelf.

> I would not recommend anyone to spend time on this until there is a script reproducing the issue.
> 
> Right now, there's no proof there's a problem.

The issue is that the build is not independently bit-by-bit reproducible. You can reproduce it with `nix-build '<nixpkgs>' -A python313 --check --keep-failed`. I don't understand what more you're looking for?

**Comment by raboof** [comment]
> I'll test with parallelism 2 just to be sure.

I can confirm the problem persists with cores/parallelism set to 2

**Comment by hacscred** [comment]
To rule it out completely, don't you need to run without any parallelism (although I think this won't lead to the root cause)? 

It's likely the build ID is created by default, although one can also disable it or explicitly enable it. 

Flags to gcc (-Wl), but likely different syntax when using the Nix build derivations will allow you to disable the build ID, which will at least eliminate the ID. It's probably possible to enable the build ID again later.

**Comment by raboof** [comment]
Looking closer the different NT_GNU_BUILD_ID might be caused by a more low-level inconsistency after all - specifically the intermediate build components `Programs/_bootstrap_python.o` and `Python/frozen.o` don't appear to be deterministic. It seems those embed compiled Python code generated by `Tools/build/freeze_modules.py --frozen-modules`. So this might be a 'python-level' nondeterminism after all, possibly https://github.com/python/cpython/issues/124442 (I was testing with 3.13.0rc2 and that fix is in 3.13.0rc3)

**Comment by raboof** [comment]
jup looks like this has been fixed with #345809 :tada:
